### PR TITLE
[Distributed] Augment accessor to be able to decode arguments instead of accepting buffer

### PR DESCRIPTION
**Pull request by xedin**
Remove `HeterogeneousBuffer` and adjust distributed target accessor to:

- Accept an instance of a concrete `InvocationDecoder` based on the distributed actor
- Call `decodeNextArgument` method on the provided decoder to retrieve argument values
- Handle errors raised by decoding
- Change `InvocationDecoder` requirement to be class bound

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

File: lib/IRGen/GenDistributed.cpp [hunk 1]
```diff
@@ -93,33 +103,17 @@ class DistributedAccessor {
...
-               SILType paramType,
-               ParameterConvention convention,
-               Explosion &arguments);
-
-  /// Load an argument value from the given buffer \c argumentSlot
+  /// Load an argument value from the given decoder \c decoder
   /// to the given explosion \c arguments. Information describing
   /// the type of argument comes from runtime metadata.
   ///
   /// Returns a pair of aligned offset and value size.
```
Comment (ktoso) [inline lib/IRGen/GenDistributed.cpp:113]: comment to be removed? we dont return those

File: test/Distributed/Runtime/distributed_actor_remoteCall.swift [hunk 2]
```diff
@@ -182,17 +189,45 @@ struct FakeInvocation: DistributedTargetInvocationEncoder, DistributedTargetInvo
...
-  // === Receiving / decoding -------------------------------------------------
+  // For testing only
+  func makeDecoder() -> FakeInvocationDecoder {
+    return .init(
+      args: arguments,
+      substitutions: substitutions,
+      returnType: returnType,
+      errorType: errorType
+    )
+  }
```
Comment (ktoso) [inline test/Distributed/Runtime/distributed_actor_remoteCall.swift:200]: ok fine workaround 👍

File: test/Distributed/Runtime/distributed_actor_remoteCall.swift [hunk 3]
```diff
@@ -403,6 +451,36 @@ func test() async throws {
...
+  try decodeErrInvocation.doneRecording()
+
+  let decodeErrDecoder = decodeErrInvocation.makeDecoder()
+  try await system.executeDistributedTarget(
+    on: local,
+    mangledTargetName: expectsDecodeErrorName,
+    invocationDecoder: decodeErrDecoder,
+    handler: FakeResultHandler()
+  )
+  // CHECK: ERROR: ExecuteDistributedTargetError(message: "Failed to decode of Int??? (for a test)")
```
Comment (ktoso) [inline test/Distributed/Runtime/distributed_actor_remoteCall.swift:482]: ❤️

File: test/Distributed/SIL/distributed_actor_remoteCall_sil.swift [hunk 4]
```diff
@@ -15,91 +15,94 @@ distributed actor MyDistActor {
 
 // TODO(distributed): very naive test, SIL for discussion and cleanup
 
-// CHECK: // MyDistActor.test()
+// MyDistActor.test()
```
Comment (ktoso) [inline test/Distributed/SIL/distributed_actor_remoteCall_sil.swift:18]: Sorry that the test is so annoying, I will make it proper as the TODO mentions above

**Review by ktoso — APPROVED** [review]
Read through and looks great, I'll work on the next SIL thing (emitting the func ref for the record) next. Thanks!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

File: include/swift/AST/TypeCheckRequests.h [hunk 5]
```diff
@@ -1091,6 +1091,43 @@ class GetDistributedActorIDPropertyRequest :
...
+  friend SimpleRequest;
+
+  NominalTypeDecl *evaluate(Evaluator &evaluator, NominalTypeDecl *actor) const;
+
+public:
+  // Caching
+  bool isCached() const { return true; }
+};
+
+/// Obtain the method that can use bed to decode argument values passed
```
[the remark this answers is not in the document]
Comment (xedin) [reply include/swift/AST/TypeCheckRequests.h:1112]: Yeah, there is an extraneous `d` in there, I'm going to fix it separately...

File: include/swift/AST/ASTContext.h [hunk 6]
```diff
@@ -1345,6 +1345,14 @@ class ASTContext final {
   /// alternative specified via the -entry-point-function-name frontend flag.
   std::string getEntryPointFunctionName() const;
 
+  /// Find the concrete invocation decoder assoicated with the given actor.
```
[the remark this answers is not in the document]
Comment (xedin) [reply include/swift/AST/ASTContext.h]: Fixed, thanks!

File: include/swift/AST/TypeCheckRequests.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (xedin) [reply include/swift/AST/TypeCheckRequests.h]: Fixed!

**Comment by xedin** [comment]
@swift-ci please test

**Pull request merged**